server/src/main/java/org/elasticsearch/transport/ConnectionManager.java

```diff
@@ -166,44 +155,28 @@ public void connectToNode(DiscoveryNode node, ConnectionProfile connectionProfil
                                     connectionListener.onNodeDisconnected(node);
                                 }));
                             }
-                            if (conn.isClosed()) {
-                                throw new NodeNotConnectedException(node, "connection concurrently closed");
-                            }
-                            success = true;
-                            listeners = connectingNodes.remove(node);
                         }
-                    } catch (ConnectTransportException e) {
-                        throw e;
-                    } catch (Exception e) {
-                        throw new ConnectTransportException(node, "general node connection failure", e);
                     } finally {
-                        if (success == false) { // close the connection if there is a failure
-                            logger.trace(() -> new ParameterizedMessage("failed to connect to [{}], cleaning dangling connections", node));
-                            IOUtils.closeWhileHandlingException(conn);
-                        } else {
-                            releaseOnce.run();
-                            ActionListener.onResponse(listeners, null);
-                        }
+                        CompletableContext<Void> completableContext = pendingConnections.remove(node);
+                        releaseOnce.run();
+                        completableContext.complete(null);
                     }
                 }, e -> {
                     assert Transports.assertNotTransportThread("connection validator failure");
-                    IOUtils.closeWhileHandlingException(conn);
-                    final List<ActionListener<Void>> listeners;
-                    try (Releasable ignored = connectionLock.acquire(node.getId())) {
-                        listeners = connectingNodes.remove(node);
+                    try {
+                        IOUtils.closeWhileHandlingException(conn);
+                    } finally {
+                        CompletableContext<Void> completableContext = pendingConnections.remove(node);
+                        releaseOnce.run();
+                        completableContext.completeExceptionally(e);
```

Comment:
are we guaranteed that `completableContext` is non-null here?

Reply:
Note that this looks very similar to the code below, so perhaps it can be shared.